### ISSUE TITLE
DEFAULTED_ON_CREATE not supported in alternate testing DB?

### ISSUE THREAD

**Issue by thatguysimon** [django-salesforce/django-salesforce#213]
Hi!
I've successfully set up an alternate testing DB with SQLite as instructed in the readme, however when running my tests, I get the following error:
```
django.db.utils.InterfaceError: Error binding parameter 1 - probably unsupported type
```

This appears to come from a field I have defined as:
```ruby
is_active = models.BooleanField(custom=True,
                                db_column='Is_Active__c', 
                                default=models.DEFAULTED_ON_CREATE)
```

A quick look at the `execute` method in `django/db/backends/sqlite3/base.py` reveals that the `Is_Active__c` param that django is trying to save is:
`<salesforce.backend.operations.DefaultedOnCreate object at 0x10f233c50>` ... instead of True/False. Changing that param value to either True or False while debugging does indeed solve the problem.

Is there a known issue working with DEFAULTED_ON_CREATE in alternate DBs? Am I missing some kind of configuration?

Thanks!

**Comment by hynekcer** [comment]
A similar more complicated issue with primary keys #231 has been fixed so that any database can be used with less restrictions not only for test but also for copy and use it as a live backup of Salesforce. There is no difference between run-time and test, but a difference between Salesforce and other database backends. The problem was that Field methods receive either `connection` and value or `instance` and value, but no method gets both information. If the value is None than these methods of SomeField are skipped.

A general solution of DEFAULTED_ON_CREATE is more complicated than the two default values. The value is not known until the record is refreshed from the database. If the value can be set to None later, then it shouldn't be set again to the default value. Your examples for the most frequent problems are fortunately easier and it works for you.

One my objection is that people probably need to test a part also on Salesforce db. The condition `settings.ENV == 'test'` is more related to the backend than to test versus run-time. Another problem of your approach is that `makemigrations` would also depend on the current value of ENV variable. That will be a mess and not for everybody.

Two solutions are possible:

A) Added an empty customizable method `.before_save(self, value, is_salesforce, is_creating, connection, instance)` to the base `SfField` class. That method will be called automatically before `save` and `bulk_create` to solve null values. Important parameters are `value, is_salesforce, is_creating` to simplify the logic, but the original full parameters `connection, instance` can be are also available. Terrible solution?

B) It seems easier to emulate these simple cases similarly to Salesforce at the database level:

    CREATE TABLE ...
    IsActive BOOLEAN DEFAULT(TRUE),
    CreatedDate TIMESTAMP  DEFAULT CURRENT_TIMESTAMP,

This will be created by `makemigrations` and internal `migrate` from

    is_active = BooleanField(...default=models.DEFAULTED_ON_CREATE('TRUE'))
    created_date = DateTimeField(default=models.DEFAULTED_ON_CREATE('CURRENT_TIMESTAMP'), 
                                 sf_read_only=models.READ_ONLY)

Both require come non trivial code, but B) can be easier explained how to use it.

**Comment by hynekcer** [comment]
The right solution should be simpler then A), B) or your code.

The algorithm of `inspectdb` command was stable long time and it ignored default values that were useful only with non-Salesforce databases.

I wrote a patch for `inspectdb` 

* Support for introspection with DefaultedOnCreate(value).  100cc68

to export more information, e.g.
``` python
    default=models.DefaultedOnCreate(False)`,
    default=models.DefaultedOnCreate('NotSeen'), choices=[('NotSeen', 'Not Seen'), ...],
```
Nothing is changed in the backend implementation yet, but you can still simplify your code by:

``` python
class TestableBooleanField(models.BooleanField):
    def __init__(self, **kwargs):
        if settings.ENV == 'test' and isinstance(kwargs['default']), DefaultedOnCreate):
            kwargs['default'] = kwargs['default'].arg
        super(TestableBooleanField, self).__init__(default=default, **kwargs)
```

(not fixed, but a progress)

* A problem is also in Salesforce with objects that were just created if they are saved once more without previous refreshing fields that were assigned by Salesforce defaults. We can try a special request that writes some fields and reads some fields in the same request. Useful for those fields with `sf_read_only` attribute (e.g. a formula that depends on ohter fields) or with `DefaultedOnCreate`

* No field on the Salesforce platform expects a default value from the application.

**Comment by hynekcer** [comment]
Most problems are already solved by the [PR for version 1.0](https://github.com/django-salesforce/django-salesforce/pull/255).

Instead of your proposal `testing_default` a parameter for DefaultedOnCreate() is implemented, e.g.:
``` python
from django.utils.timezone import now

class ...:
    is_active = models.BooleanField(default=models.DefaultedOnCreate(True))
    created_date = models.DateTimeField(default=models.DefaultedOnCreate(now),
                                        sf_read_only=models.READ_ONLY)
```
A documentation for it is in wiki [Introspection and Special Attributes of Fields](https://github.com/django-salesforce/django-salesforce/wiki/Introspection-and-Special-Attributes-of-Fields).

Missing part: tests (a new test project that runs on Salesforce then on 'default' and possibly also with and without migrations)
